OMCompiler/Compiler/Template/CodegenFMUCommon.tpl

```diff
@@ -659,12 +659,63 @@ template UnitDefinitions(SimCode simCode)
  "Generates code for UnitDefinitions file for FMU target."
 ::=
 match simCode
-case SIMCODE(__) then
+case SIMCODE(modelInfo=modelInfo) then
+match modelInfo
+case MODELINFO(unitDefinitions = unitDefinitions) then
+  <<
+  <%UnitDefinitionsHelper(unitDefinitions)%>
+  >>
+end UnitDefinitions;
+
+template UnitDefinitionsHelper(list<UnitDefinition> unitDefinitions)
+ "Generates code for UnitDefinition for FMU target."
+::=
+  if unitDefinitions then
   <<
   <UnitDefinitions>
+    <%unitDefinitions |> unitDefinition => UnitDefinitionsHelper1(unitDefinition) ;separator="\n"%>
   </UnitDefinitions>
   >>
-end UnitDefinitions;
+end UnitDefinitionsHelper;
+
+template UnitDefinitionsHelper1(UnitDefinition unitDefinition)
+ "helper function to generates code for UnitDefinition for FMU target."
+::=
+match unitDefinition
+case UNITDEFINITION(name = name, baseUnit = baseUnit) then
+  <<
+  <Unit <%unitDefinitionAttribute(name)%>>
+    <BaseUnit <%baseUnitAttributes(baseUnit)%>/>
+  </Unit>
+  >>
+end UnitDefinitionsHelper1;
+
+template unitDefinitionAttribute(String unitName)
+ "Generates code for UnitDefinition Attribute for FMU target."
+::=
+  let unitString = if unitName then 'name ="<%unitName%>"'
+  <<
+  <%unitString%>
+  >>
+end unitDefinitionAttribute;
+
+template baseUnitAttributes(BaseUnit baseUnit)
+ "Generates code for BaseUnit for FMU target."
+::=
+match baseUnit
+case (BASEUNIT(mol = mol, cd = cd, m = m, s = s, A = A, K = K, kg = kg, factor = factor, offset = offset)) then
+  let mol_Value = if not intEq(mol, 0) then ' mol="<% mol %>"' else ""
+  let cd_Value = if not intEq(cd, 0) then ' cd="<% cd %>"' else ""
+  let m_Value = if not intEq(m, 0) then ' m="<% m %>"' else ""
+  let s_Value = if not intEq(s, 0) then ' s="<% s %>"' else ""
+  let A_Value = if not intEq(A, 0) then ' A="<% A %>"' else ""
+  let K_Value = if not intEq(K, 0) then ' K="<% K %>"' else ""
+  let kg_Value = if not intEq(kg, 0) then ' kg="<% kg %>"' else ""
+  let factor_Value = if not intEq(realInt(factor), 1) then ' factor="<% factor %>"' else ""
```

Comment:
factor is not an integer

Reply:
Please add offset as well

Reply:
Yes factor is not an integer, only real values are populated, the intEq is just used for comparison,  i am unable to use realEq in .tpl files

Reply:
But you cannot use intEq to compare reals

Reply:
i am converting real to int using realInt and then comparing

Reply:
> Please add offset as well

added offset Value